scripts/test_solidity_lsp.py

```diff
@@ -0,0 +1,379 @@
+#!/usr/bin/env python3.8
+
+import argparse
+import json
+import os
+import subprocess
+
+from typing import List, Union, Any
+from deepdiff import DeepDiff
+
+# {{{ JsonRpcProcess
+class MyEncoder(json.JSONEncoder):
+    """
+    Encodes an object in JSON
+    """
+    def default(self, o): # pylint: disable=E0202
+        return o.__dict__
+
+class JsonRpcProcess:
+    exe_path: str
+    exe_args: List[str]
+    process: subprocess.Popen
+    trace_io: bool
+
+    def __init__(self, exe_path: str, exe_args: List[str], trace_io: bool = False):
+        self.exe_path = exe_path
+        self.exe_args = exe_args
+        self.trace_io = trace_io
+
+    def __enter__(self):
+        self.process = subprocess.Popen(
+            [self.exe_path, *self.exe_args],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE
+        )
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback) -> None:
+        self.process.kill()
+        self.process.wait(timeout=2.0)
+
+    def trace(self, topic: str, message: str) -> None:
+        if self.trace_io:
+            print(f"{SGR_TRACE}{topic}:{SGR_RESET} {message}")
+
+    def receive_message(self) -> Any:
+        # `, timeout: float = 2.0`
+        # Note, we should make use of timeout to avoid infinite blocking if nothing is received.
+        LEN_HEADER = "Content-Length: "
+        TYPE_HEADER = "Content-Type: "
+        if self.process.stdout == None:
+            return None
+        message_size = None
+        while True:
+            #read header
+            line = self.process.stdout.readline()
+            if not line:
+                # server quit
+                return None
```

Comment:
I think that you should be able to just iterate over the output with `readlines()` instead of `readline()`:
```suggestion
        # iterate over all headers
        for line in self.process.stdout.readlines():
            # <rest of the loop body here>
        else:
            # server quit
            return None
```

Reply:
readlines() **will** hang until `stdout` is being fully read, which is not the case, I want to read interactively until an empty line has been received. I don't think this won't work. (Also, I had to google up the for/else clause first, it wasn't intuitive to me what that could mean :) )

Reply:
> readlines() will hang until stdout is being fully read

ok, that's a good argument against it.

I just think it would be better to have a `for` loop. These `while True` loops are harder to read for me because I don't immediately see what they're iterating over. But if there are problems with that then I guess we have to stay with a `while` loop.